### DISCUSSION TITLE
`&JsValue` -> `&T` where `T` is an exported Rust struct

### DISCUSSION THREAD

**Discussion by WilliamVenner** [wasm-bindgen/wasm-bindgen#3943] [Q&A] [answered]
Hi, is there a way to convert a `&JsValue` to a `&T` where `T` is an exported Rust struct?

The `TryFromJsValue` trait consumes a `JsValue` and even frees the underlying exported Rust struct from the heap.

`JsCast` isn't implemented for exported Rust structs.

**Comment by Liamolucko** [comment, marked as answer]
No, there isn't at the moment.

A literal `&JsValue` -> `&T` transformation wouldn't work. The reason is that the way that borrowing Rust types works is that `wasm-bindgen` wraps them in an internal `RefCell` which it borrows every time a reference is passed to an exported Rust function, and then at the end of the function the `Ref` / `RefMut` is dropped to mark the `RefCell` as non-borrowed again. Otherwise JavaScript could easily pass the same `T` in for two different `&mut T` parameters of a function and cause undefined behaviour.

With a `&T`, there's no way to run some code when it's dropped; so, such an API would require creating some kind of `Ref<T>` or `Handle<T>` type which you can then turn `&JsValue` into.

---

Also, `TryFromJsValue` is a private trait (for now, anyway) - it's in `wasm_bindgen::convert`, the docs for which say:

> This is mostly an internal module, no stability guarantees are provided. Use at your own risk.

It probably should be public, but it's a bit of a mess: see #3709 for the history here.

**Reply by WilliamVenner** [reply]
Thank you!
